src/blr/cli/clusterrmdup.py

```diff
@@ -3,594 +3,303 @@
+class CacheReadPairTracker():
     """
-    Seeds duplicates for read pairs
-    :param duplicate_position_dict: Dictionary with read pairs where both read&mate are marked as duplicates
-    :return: Nothing, sends results to possible_duplicate_seeds & overlapValues
+    Stores read pairs and keeps track is reads/mates are marked as duplicate for that set of reads.
     """
 
-    # If all reads for this chromosome has been "unpaired duplicates", won't be able to seed => return
-    if not chromosome in duplicate_position_dict:
-        return
-
-    # Function takes long time, so nice to see if something happens
-    # UPDATE
-    progressBar = BLR.ProgressBar(name=str('Seeding ' + str(chromosome)), min=0, max=len(duplicate_position_dict[chromosome]), step=1)
+    def __init__(self, rp_pos_tuple, chromosome, read, mate, bc):
 
-    for duplicate_pos in sorted(duplicate_position_dict[chromosome].keys()):
-
-        # Devide into exactly matching positions (rp1_pos == rp2_pos == rp3_pos...)
-        possible_duplicate_seeds = dict()
-        for readpair in duplicate_position_dict[chromosome][duplicate_pos]:
-
-            # Fetching information from read/mate
-            mate = readpair[0]
-            read = readpair[1]
-            read_pos_tuple = (read.get_reference_positions()[0], read.get_reference_positions()[-1])
-            mate_pos_tuple = (mate.get_reference_positions()[0], mate.get_reference_positions()[-1])
-            readpair_pos_tuple = (mate_pos_tuple, read_pos_tuple)
-            try:barcode_ID = int(read.get_tag(barcode_tag))
-            except KeyError:
-                if not force_run:
-                    import sys
-                    sys.exit('\nERROR: No BC tag for read ' + str(read.query_name) + '\nPlease double check BC '
-                             'clustering & tagging has run correctly.\n\nIf BC indexing nucleotides contain N, '
-                             'either remove all affected reads or cluster these sequences separately and append '
-                             'to the end of your BC.N.clstr file.\nEXITING')
+        self.position_tuple_ID = rp_pos_tuple
+        self.chromosome = chromosome
+        self.current_reads = dict()
+        self.barcodes = set()
+        self.read_pos_has_duplicates = bool()
+        self.mate_pos_has_duplciates = bool()
 
-            # Add all barcodes IDs to set, check later if total > 2 at positions.
-            if not readpair_pos_tuple in possible_duplicate_seeds:
-                possible_duplicate_seeds[readpair_pos_tuple] = set()
-            possible_duplicate_seeds[readpair_pos_tuple].add(barcode_ID)
+        if read and mate:
+            self.add_read_pair(read=read, mate=mate, bc=bc)
 
-        # When all overlaps found for current position, try finding proximal read pairs with same overlap
-        for possible_seed, barcode_IDs in possible_duplicate_seeds.items():  # Only one entry
+    def add_read_pair(self, read, mate, bc):
```

Comment:
On this point I am not sure if I agree. 

The read pair cache in its most basic component only tracks the read pairs for any given `read_pos_tuple` and it is only for convenience it has other functions and such build into it, such as a `.barcodes` variable (instead of calling on the barcodes in a for-loop over the read pairs). It was also a way to avoid using the `.get_tag()` function multiple times since bc is already fetched by prior to this step.

Although I see your point that it might be confusing and I guess the function does not need to track the reads itself now that it has a quite clear cut purpose as it would suffice to only track the relevant information about the read (in this case `rp_pos_tuple`, `barcodes`, `read_duplicate` and `mate_duplicate`).

In that case a more suitable name would maybe be something along the line of `barcode_set_cache`, which then has an ID according `rp_pos_tuple` and stores whether it has duplicates at both read/mate positions using those variables. What do you think of that?